### ISSUE TITLE
AWD LSTM dropout makes 'weight_hh_l0' a leaf Parameter rather a non-leaf Tensor

### ISSUE THREAD

**Issue by Zhong-Zhang** [fastai/fastai#2591]
In file **awd_lstm.py**, at class **WeightDropout**, there is a code in functions **__init__** and **reset**:
`
self.module._parameters[layer] = F.dropout(raw_w, p=self.weight_p, training=False)
`

The 'self.module._parameters[layer]' here is typically LSTM's Parameter 'weight_hh_l0'.

With training=False, F.dropout doesn't do anything, which makes the output identical to 'raw_w', i.e., a shallow copy. The problem is, raw_w is a Parameter (leaf), but from my understanding, 'weight_hh_l0' should be an intermediate Tensor (non-leaf) which is computed from 'raw_w'.

This could cause side effects, e.g., one might accidentally put non-leaf variables if calling model.parameters() into the optimizer or get error from function clip_grad_norm_.

Here is my patch for this issue, I'm willing to do a PR if I'm correct.

`def __init__`
```
            # self.module._parameters[layer] = F.dropout(w, p=self.weight_p, training=False)
            del self.module._parameters[layer]
            setattr(self.module, layer, self._mask_dropout(w, self.weight_p, training=False)) # make it a tensor not parameter
```

`def _setweights`
```
            # self.module._parameters[layer] = F.dropout(raw_w, p=self.weight_p, training=self.training)
            # if self.idxs is not None: self.module._flat_weights[self.idxs[i]] = self.module._parameters[layer]
            setattr(self.module, layer, self._mask_dropout(raw_w, self.weight_p, training=self.training))
            if self.idxs is not None: self.module._flat_weights[self.idxs[i]] = getattr(self.module, layer)
```

`def reset`
```
            # self.module._parameters[layer] = F.dropout(raw_w, p=self.weight_p, training=False)
            setattr(self.module, layer, self._mask_dropout(raw_w, self.weight_p, training=False))
```

`def _mask_dropout`
```
    @staticmethod
    def _mask_dropout(x, drop_p, training=True):
        mask = x.new_ones(x.size())
        mask = F.dropout(mask, p=drop_p, training=training)
        x = x * mask
        return x
```

**Comment by Zhong-Zhang** [comment]
I notice that there seem issues in fastai2 as well, though I didn't dig into that too much.

**Comment by jph00** [comment]
This should be fixed in fastai v1 now. Feel free to do a PR to v2 with something similar, if you think it looks reasonable.

**Comment by jph00** [comment]
Sorry for the delay on this one. In fastai2 can you take a look at how I've handled it in `32_text.models.awdlstm.ipynb`? When I check `named_parameters` it looks OK to me. Please let me know if you see any problems. If you think it's OK, I'd be happy to accept a PR that does something similar for fastai v1.

**Comment by Zhong-Zhang** [comment]
> Sorry for the delay on this one. In fastai2 can you take a look at how I've handled it in `32_text.models.awdlstm.ipynb`? When I check `named_parameters` it looks OK to me. Please let me know if you see any problems. If you think it's OK, I'd be happy to accept a PR that does something similar for fastai v1.

Thank you for your reply. I took a look at `32_text.models.awdlstm.ipynb `, I think there still be a problem.

Please try my test ipynb code:
```
#export
class WeightDropout(Module):
    "A module that warps another layer in which some weights will be replaced by 0 during training."

    def __init__(self, module, weight_p, layer_names='weight_hh_l0'):
        self.module,self.weight_p,self.layer_names = module,weight_p,L(layer_names)
        for layer in self.layer_names:
            #Makes a copy of the weights of the selected layers.
            w = getattr(self.module, layer)
            delattr(self.module, layer)
            self.register_parameter(f'{layer}_raw', nn.Parameter(w.data))
            setattr(self.module, layer, w.clone())
            if isinstance(self.module, (nn.RNNBase, nn.modules.rnn.RNNBase)):
                self.module.flatten_parameters = self._do_nothing

    def _setweights(self):
        "Apply dropout to the raw weights."
        for layer in self.layer_names:
            raw_w = getattr(self, f'{layer}_raw')
            if self.training: w = F.dropout(raw_w, p=self.weight_p)
            else: w = raw_w.clone()
            setattr(self.module, layer, w)
            
            print("drop p:", self.weight_p)
            print("raw_w type, should be Parameter:", type(getattr(self, f'{layer}_raw')))
            print("w type, should be Tensor:", type(getattr(self.module, f'{layer}')))
            print("raw_w leaf, should be True:", getattr(self, f'{layer}_raw').is_leaf)
            print("w leaf, should be False:", getattr(self.module, f'{layer}').is_leaf)
            print("Is w a shallow copy of raw_w:", id(raw_w)==id(w))


    def forward(self, *args):
        self._setweights()
        with warnings.catch_warnings():
            # To avoid the warning that comes because the weights aren't flattened.
            warnings.simplefilter("ignore", category=UserWarning)
            return self.module.forward(*args)

    def reset(self):
        for layer in self.layer_names:
            raw_w = getattr(self, f'{layer}_raw')
            setattr(self.module, layer, raw_w.clone())
        if hasattr(self.module, 'reset'): self.module.reset()

    def _do_nothing(self): pass

# when p is not 0
module = nn.LSTM(5,7)
dp_module = WeightDropout(module, 0.4)
wgts = dp_module.module.weight_hh_l0
tst_inp = torch.randn(10,20,5)
h = torch.zeros(1,20,7), torch.zeros(1,20,7)
dp_module.reset()
x,h = dp_module(tst_inp,h)
loss = x.sum()
loss.backward()

print('\n')

# when p is 0
module = nn.LSTM(5,7)
dp_module = WeightDropout(module, 0.0)
wgts = dp_module.module.weight_hh_l0
tst_inp = torch.randn(10,20,5)
h = torch.zeros(1,20,7), torch.zeros(1,20,7)
dp_module.reset()
x,h = dp_module(tst_inp,h)
loss = x.sum()
loss.backward()
```

Which will produce:
```
drop p: 0.4
raw_w type, should be Parameter: <class 'torch.nn.parameter.Parameter'>
w type, should be Tensor: <class 'torch.Tensor'>
raw_w leaf, should be True: True
w leaf, should be False: False
Is w a shallow copy of raw_w: False


drop p: 0.0
raw_w type, should be Parameter: <class 'torch.nn.parameter.Parameter'>
w type, should be Tensor: <class 'torch.nn.parameter.Parameter'>
raw_w leaf, should be True: True
w leaf, should be False: True
Is w a shallow copy of raw_w: True
```

As we can see, when p!=0, everything is ok. However, when p==0, something is broken.
This problem roots in F.dropout() simply pass the input as the output, as id(raw_w)==id(w). That is why I use a function _mask_dropout.

**Comment by Zhong-Zhang** [comment]
To my understanding, this F.dropout(x, p=0) thing seems a pytorch bug (or a weird feature?).
A simple fix is to apply dropout on a mask, and then multiply it to the variable which actually needs dropout:
```
def _mask_dropout(x, drop_p, training=True):
        mask = x.new_ones(x.size())
        mask = F.dropout(mask, p=drop_p, training=training)
        x = x * mask
        return x
```
This operation forces the output be always computed from the input, not give a shallow copy. The side effect is a slight increase in memory.

I don't know if there are other places have the same problem. However, if it does, this simple fix seems not so good, should we need to consider a more comprehensive plan about this?

**Comment by Zhong-Zhang** [comment]
I go to the pytorch dropout implement:
https://github.com/pytorch/pytorch/blob/5d474e1812b2343b7dc1c9561f5f154334ccae38/torch/csrc/api/src/nn/modules/dropout.cpp#L24-L34
```
Tensor DropoutImplBase<Derived>::forward(Tensor input) {
  if (options.rate_ == 0 || !this->is_training()) {
    return input;
  }

  auto scale = 1.0f / (1.0f - options.rate_);
  auto boolean_mask = noise_mask(input).uniform_(0, 1) > options.rate_;
  auto noise = boolean_mask.to(input.dtype()).mul_(scale);

  return input * noise;
}
```
It also uses a mask to implement dropout, but let ouput=input when p==0. Maybe we can write a similar function but remove `options.rate_ == 0`.
